### PR TITLE
[Distributed] Split data to ensure data locality for multiple clients

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
The previous version can only works for 1 client per machine. This one splits the workset further for multiple clients in a machine. It's common to have multiple clients on one machine. e.g., we run one client for each GPU or for each NUMA node.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -490,7 +497,12 @@ def rank(self):
         int
             The rank of the current graph store.
         '''
-        return self._client.client_id
+        if self._gpb is None:
+            return self._client.client_id
+        else:
+            num_client_per_part = self._num_client // self._gpb.num_partitions()
+            client_id_in_part = self._client.client_id % num_client_per_part
```
Comment (zheng-da) [inline python/dgl/distributed/dist_graph.py:504]: @aksnzhy how is the client_id decided in kvstore client?

Comment (aksnzhy) [reply python/dgl/distributed/dist_graph.py:510]: The code is right. We assume that the client id is grouped together inside a machine.

File: python/dgl/distributed/dist_graph.py [hunk 2]
```diff
@@ -494,10 +495,20 @@ def rank(self):
...
         if self._g is None:
-            return self._client.client_id
+            return rpc.get_rank()
         else:
-            return self._gpb.partid
+            # If DistGraph has a local partition, we should be careful about the rank
+            # we return. We need to return a rank that node_split or edge_split can split
+            # the workload with respect to data locality.
+            num_client = rpc.get_num_client()
+            num_client_per_part = num_client // self._gpb.num_partitions()
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:508]: Do we have the case that the number of clients can be different on each machine?

File: python/dgl/distributed/dist_graph.py [hunk 3]
```diff
@@ -558,7 +569,26 @@ def _get_overlap(mask_arr, ids):
         masks = F.gather_row(mask_arr.tousertensor(), ids)
         return F.boolean_mask(ids, masks)
 
-def node_split(nodes, partition_book, rank):
+def _split_local(partition_book, rank, elements, local_eles):
+    ''' Split the input element list with respect to data locality.
+    '''
+    num_clients = rpc.get_num_client()
+    num_client_per_part = num_clients // partition_book.num_partitions()
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:576]: same

File: python/dgl/distributed/dist_graph.py (hunk 2 above)
Comment (aksnzhy) [reply python/dgl/distributed/dist_graph.py:508]: If we will meet this case in the future, we can add a new API `get_num_client_per_machine(machine_id)`

Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:508]: it sounds a very complex scenario. why do we want to run different numbers of clients on different machines?

**Pull request merged**